plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx

```diff
@@ -0,0 +1,67 @@
+/**
+ * External dependencies
+ */
+import { useEffect } from '@wordpress/element';
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+import { BlockEditProps } from '@wordpress/blocks';
+import { getSettingWithCoercion } from '@woocommerce/settings';
+import { isBoolean } from '@woocommerce/types';
+
+/**
+ * Internal dependencies
+ */
+import { useIsDescendentOfSingleProductBlock } from '../../atomic/blocks/product-elements/shared/use-is-descendent-of-single-product-block';
+import { AddToCartOptionsSettings } from './settings';
+import { INNER_BLOCKS_TEMPLATE } from './constants';
+export interface Attributes {
+	className?: string;
+	isDescendentOfSingleProductBlock: boolean;
+}
+
+export type FeaturesKeys = 'isBlockifiedAddToCart';
+
+export type FeaturesProps = {
+	[ key in FeaturesKeys ]?: boolean;
+};
+
+export type UpdateFeaturesType = ( key: FeaturesKeys, value: boolean ) => void;
+
+// Pick the value of the "blockify add to cart flag"
+const isBlockifiedAddToCart = getSettingWithCoercion(
+	'isBlockifiedAddToCart',
+	false,
+	isBoolean
+);
+
+const AddToCartOptionsEdit = ( props: BlockEditProps< Attributes > ) => {
+	const { setAttributes } = props;
+
+	const blockProps = useBlockProps( {
+		className: 'wc-block-add-to-cart-with-options-wrapper',
+	} );
+	const { isDescendentOfSingleProductBlock } =
+		useIsDescendentOfSingleProductBlock( {
+			blockClientId: blockProps?.id,
+		} );
+
+	useEffect( () => {
+		setAttributes( {
+			isDescendentOfSingleProductBlock,
+		} );
+	}, [ setAttributes, isDescendentOfSingleProductBlock ] );
```

Comment:
I know this comes from the previous block, but I think setting attributes after the first render should be considered an anti-pattern. It causes some weird bugs in the Site Editor (https://github.com/woocommerce/woocommerce/issues/48936) and it also makes the usage of patterns problematic, as a pattern could be used in both contexts.

So I don't think we should block this PR because of this, but I wonder if we should have a follow-up issue to make sure we refactor this before releasing. What do you think?

Reply:
Maybe we can try to find a different way to handle this instead of using a block attribute.
The main reason is that we need to detect it on the server side, right?

Reply:
Yeah, that's my understanding as well. I don't know yet which alternative approach we could try, but I would love to find another way. :smile:

Reply:
I just created the issue [53283](https://github.com/woocommerce/woocommerce/issues/53283) to tackle this problem.